### PR TITLE
[20240805] 정우용_10회차 문제 풀이 6주차

### PR DESCRIPTION
**Pull request by wooyong99**
### 문제
- 스타트와 링크 : https://www.acmicpc.net/problem/14889
- 부분수열의 합 : https://www.acmicpc.net/problem/1182
- 차이를 최대로 : https://www.acmicpc.net/problem/10819
- 모든 순열 : https://www.acmicpc.net/problem/10974
- 외판원 순회 2 : https://www.acmicpc.net/problem/10971
- 부등호 : https://www.acmicpc.net/problem/2529
- 도영이가 만든 맛있는 음식 : https://www.acmicpc.net/problem/2961

### CODE REVIEW

File: [10회차] 백트래킹/우용/S2_2961.java
Comment (ChangHee98) [inline [10회차] 백트래킹/우용/S2_2961.java:1]:
```
package day10;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class S2_2961 {
    static int[] sour;
    static int[] bitter;
    static int n;
    static int min = Integer.MAX_VALUE;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        sour = new int[n];
        bitter = new int[n];
        for(int i=0; i<n; i++){
            StringTokenizer st = new StringTokenizer(br.readLine());
            sour[i] = Integer.parseInt(st.nextToken());
            bitter[i] = Integer.parseInt(st.nextToken());
        }
        boolean[] visited = new boolean[n];

        backtrack(0, 1, 0, visited);
        System.out.println(min);
    }
    public static void backtrack(int depth, int sum_sour, int sum_bitter, boolean[] visited){
        if(depth <= n && depth > 0){
            if(min > Math.abs(sum_sour - sum_bitter)){
                min = Math.abs(sum_sour - sum_bitter);
            }
        }

        for(int i=0; i<n; i++){
            if(!visited[i]){
                visited[i] = true;
                sum_sour *= sour[i];
                sum_bitter += bitter[i];
                backtrack(depth+1, sum_sour, sum_bitter, visited);
                visited[i] = false;
                sum_sour /= sour[i];
                sum_bitter -= bitter[i];
            }
        }
    }
}

```

Comment (koosco) [reply [10회차] 백트래킹/우용/S2_2961.java:1]: 도영이컷
```
import sys
from typing import List

read = sys.stdin.readline


def sol(s: int = 1, b: int = 0, idx: int = 0, cnt: int = 0):
    if idx == n:
        if not cnt:
            return
        global res
        res = min(res, abs(s - b))
        return

    sol(s * ingredients[idx][0], b + ingredients[idx][1], idx + 1, cnt + 1)
    sol(s, b, idx + 1, cnt)


n = int(read())
ingredients = [list(map(int, read().split())) for _ in range(n)]
res = sys.maxsize
sol()
print(res)
```

Comment (koosco) [reply [10회차] 백트래킹/우용/S2_2961.java:1]: 부등호
```
def sol(path: str, idx: int = 1, depth: int = 0):
    if depth == k:
        if eval(path):
            res.append(''.join(path[::2]).zfill(k + 1))
        return
    if path != '0' and not eval(path):
        return
    for i in range(10):
        if not visited[i]:
            visited[i] = True
            sol(path + ops[idx] + str(i), idx + 1, depth + 1)
            visited[i] = False


k = int(input())
ops = [''] + input().split()
visited = [False] * 10
res = []
for i in range(10):
    visited[i] = True
    sol(str(i))
    visited[i] = False
print(res[-1], res[0], sep='\n')

```

**Comment by MaengSung** [comment]
도영이 컷컷컷

N = int(input())

base = []
for _ in range(N):
    base.append(tuple(map(int,input().split())))

bb = []
sb=[]
def col(idx):
    if idx == N:
        return
    for i in range(idx,N):
        sb.append(base[i])
        bb.append(sb.copy())
        col(i+1)
        sb.pop()

def cal():
    minn = int(1e9)
    for i in bb:
        ts = 1
        tb = 0
        for j in i:
            s,b = j[0],j[1]
            ts *= s
            tb += b
        minn = min(minn,abs(ts-tb))
    return minn

col(0)
print(cal())

**Comment by MaengSung** [comment]
부등호

def sol(path:str,idx:int=0, depth:int=0):
    if depth == k:
        if eval(path):
            res.append(''.join(path[::2]).zfill(k+1))
        return
    if path != '0' and not eval(path):
        return
    for i in range(10):
        if not visited[i]:
            visited[i] = True
            sol(path+ops[idx]+str(i),idx+1,depth+1)
            visited[i] = False

k = int(input())
ops = []+input().split()
visited = [False] * 10
res = []
for i in range(10):
    visited[i] = True
    sol(str(i))
    visited[i] = False
print(res[-1],res[0],sep='\n')

**Pull request merged**